### ISSUE TITLE
Can WebGPU canvas alpha be configured?

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1425]
For WebGPU (gpupresent) canvases, I want to propose that the following options (from WebGL) would be **unconfigurable** and fixed at the following values. I thought we previously agreed on these, but I'm not sure where, and it would have been a VERY long time ago.

- `alpha: true` - In WebGL, alpha:false has significant cost on some platforms because it's necessary to use RGBA instead of RGB, meaning the platform surface/overlay's alpha channel needs to get cleared or guarded against writes. It's possible that we can sidestep this issue in WebGL - perhaps problems that force us to use RGBA won't be present, but I don't think we should bank on it. This may also change the swap chain format to e.g. rgb8unorm, which we don't even have.
- `premultipliedAlpha: true` - Platform surfaces/overlays generally require premultiplied alpha (AFAIK). Unfortunately, this means color values where `r>a || g>a || b>a` cause undefined compositing results (this is true in WebGL) (but not undefined JS-visible results).

Related: https://github.com/gpuweb/gpuweb/pull/1408#discussion_r570394799 for `preserveDrawingBuffer`

CC cts issue: https://github.com/gpuweb/cts/issues/460

**Comment by greggman** [comment]
I clearly don't know all the issues but for me alpha:false is supposed = speed, blending is slower than not blending, alpha:false = not blending. Maybe on a 3080 it's a tiny speed difference but on the slowest Intel GPU that can support WebGPU is it?

As for premultipliedAlpha, the original point is that pretty much all native games I've ever worked on expected to be writing to an opaque background so allowing premultipliedAlpha = false, alpha = false #1 use case is for porting native apps.

Checking I see that all unity games run with premultipliedAlpha = false, alpha = true, unreal games run with premuipliedAlpha = true, alpha = false (which ends up being the same as premultipliedAlpha true because with alpha = false, alpha is effectively 1.0

Further, while this is wishful speculation, I'd always hoped there would be fast path in the browser, like if you used the fullscreen API to promote your canvas to be the fullscreen element, and premultipliedAlpha = false, and alpha = false and preserveDrawingBuffer = false, then why not draw the canvas directly to the window instead of compositing?

**Comment by juj** [comment]
I am with @greggman here that I have always been very surprised about Chrome's performance penalty in WebGL when `alpha: false` is specified.

It seems as if Chrome fuses together the use of enable-vs-disable alpha blending, and RGB-vs-RGBA ("RGBX") memory layout.

` ... or guarded against writes`

This is also something I don't understand. To my view, WebGL implementations would not need to do anything different when `alpha: false` context is created. The alpha channel can exist and sit there, and one can render to it, or not render to it; its contents would just be ignored when the browser composites the canvas with the rest of the page(?).

Not sure if there is something awkward behavior that got specced along the way for Canvas/WebGL about this, but ideally that is the way it would be fastest and cleanest to work. Just because one is not alpha blending should not mean that one should murder their memory layout.

In native APIs, disabling alpha blending and using memory layout RGBA==RGBX is the fastest option. Alpha channel is just ignored.

So I am not sure why Chrome would use slower unaligned RGB render targets when `alpha: false` is used. It should just be the blending mode that changes?

> * This may also change the swap chain format to e.g. rgb8unorm, which we don't even have.

If the canvas compositing mode was carefully specced not to affect the memory storage layout, then this would become trivial and orthogonal?

Also instead of making alpha and premultipliedAlpha gpupresent context creation modes, I would prefer that for WebGPU, `HTMLCanvasElement` gained a new property 
```
HTMLCanvasElement.webGpuCompositingMode = 'copy' | 'alpha-blend' | 'premultiplied-alpha-blend';
```
(or one could even go for the full `GPUBlendComponent` operation matrix if that is deemed useful for some effects)

so that the blending mode could be adjusted at runtime.

Then there is already the existing `GPUCanvasContext.configureSwapChain()` function, which allows one to separately specify what the backbuffer storage format is.

The browser would then do a regular blit of the swap chain texture onto the rest of the canvas, using the blend function specified on the canvas.

The reason that `webGpuCompositingMode` would be nice to be a runtime changeable property instead of a gpupresent context creation time thing is that it is awkward to require applications to know at context initialization time what kind of blend mode they will use. They might know that only later after their engine user content has been loaded in. Also that will prevent from utilizing effects at runtime, e.g. where one might want to enable slower blending at some point of the application+web page interaction, but then disable the blending operation at later stages.

Also it does not feel nice to require premultipliedAlpha everywhere. That could just as easily be the choice of the application? If PMA is required, that can cause apps to need to do an extra fullscreen non-PMA->PMA blit pass beforehand to fix the color data, which can severely impact fillrate performance on lower end Android devices (something that we see being really scarce already with existing Chrome WebGL today)

**Comment by greggman** [comment]
> so that the blending mode could be adjusted at runtime.

FYI, You can already adjust the blending mode at runtime. From the point of view of HTML, a canvas is just another element like all others. Elements use CSS and CSS has [opacity](https://developer.mozilla.org/en-US/docs/Web/CSS/opacity) as well as [a bunch of different blending modes](https://developer.mozilla.org/en-US/docs/Web/CSS/mix-blend-mode), [filtering](https://developer.mozilla.org/en-US/docs/Web/CSS/image-rendering) basically meaning being able to choose texture filtering (nearest/linear) as well as various [image processing filters](https://developer.mozilla.org/en-US/docs/Web/CSS/filter)

**Comment by greggman** [comment]
> This is also something I don't understand. To my view, WebGL implementations would not need to do anything different when alpha: false context is created. The alpha channel can exist and sit there, and one can render to it, or not render to it; its contents would just be ignored when the browser composites the canvas with the rest of the page(?).

You can't ignore it because blending can use DST_ALPHA etc.... so the results will be different if alpha exists vs not existing, regardless of whether it's used when compositing.

**Comment by juj** [comment]
> You can't ignore it because blending can use DST_ALPHA etc.... so the results will be different if alpha exists vs not existing, regardless of whether it's used when compositing.

"Ignoring" there referred to when browser is compositing the canvas with the rest of the page when `alpha: false`. There is no way to configure DST_ALPHA there.

If alpha: true/false was specced to also affect the storage of the default back buffer ("one shall not observe an alpha channel on the default backbuffer inside the WebGL application if alpha: false is set at canvas context creation"), then I see that as a silly mistake in the WebGL+canvas spec, which results in the performance problems that result in recommending `alpha: true` to be the faster option. That was the point of my post.

**Comment by kainino0x** [comment]
@kenrussell knows more about the issues chromium has encountered.

**Comment by kainino0x** [comment]
> This is also something I don't understand. To my view, WebGL implementations would not need to do anything different when `alpha: false` context is created. The alpha channel can exist and sit there, and one can render to it, or not render to it; its contents would just be ignored when the browser composites the canvas with the rest of the page(?).

I think the problem (for both alpha and premultipliedAlpha) is the browser is often _not_ responsible for compositing. Canvases often end up as overlay/underlay planes or OS compositor surfaces (IOSurfaces, DXGI surfaces, etc.) I don't know the details of what knobs we have to control the OS compositors.

> If the canvas compositing mode was carefully specced not to affect the memory storage layout, then this would become trivial and orthogonal?

👍 if possible

> The reason that `webGpuCompositingMode` would be nice to be a runtime changeable property instead of a gpupresent context creation time thing is that it is awkward to require applications to know at context initialization time what kind of blend mode they will use.

In WebGPU configureSwapChain can be done multiple times, so it could potentially be in there instead of on the canvas element.

**Comment by kenrussell** [comment]
> > This is also something I don't understand. To my view, WebGL implementations would not need to do anything different when `alpha: false` context is created. The alpha channel can exist and sit there, and one can render to it, or not render to it; its contents would just be ignored when the browser composites the canvas with the rest of the page(?).
> 
> I think the problem (for both alpha and premultipliedAlpha) is the browser is often _not_ responsible for compositing. Canvases often end up as overlay/underlay planes or OS compositor surfaces (IOSurfaces, DXGI surfaces, etc.) I don't know the details of what knobs we have to control the OS compositors.

That's correct. WebGL implementations on several platforms optimize for power consumption by handing the platform-specific surface concept (on top of which the OpenGL texture for the WebGL back buffer is layered) to the OS's compositor. In general the OS compositors do not provide a control to ignore the alpha channel of the surface. (My own experience has been on macOS and ChromeOS.) Some platforms can bind an RGB texture to the surface and write 1.0 alpha into it, but that's broken specifically on macOS, so the browser has to perform costly workarounds which amount to an additional full-canvas blit. See http://crbug.com/1045643 for more details.

Today's OS compositors also do not have an option to treat the alpha channel as not premultiplied. WebGL's `premultipliedAlpha:false` implies another full-canvas blit on some platforms.

The inevitable performance pitfalls associated with WebGL's `alpha:false` and `premultipliedAlpha:false` context creation options continue to cost engineering time, so for WebGPU I advise not providing control over these, stating that the back buffer always has an alpha channel, and that it's treated as though the alpha channel is premultiplied. I don't think undue burden would be imposed on applications desiring opaque content by requiring them to fill out the alpha channel to 1.0.

**Comment by greggman** [comment]
> Today's OS compositors also do not have an option to treat the alpha channel as not premultiplied. WebGL's premultipliedAlpha:false implies another full-canvas blit on some platforms.

I'm not following here.  Native games do not deal with this issue. How are they working if they are not supplying an alpha channel? They are not having to work around this issue at all. They just ask for an opaque window and they get an opaque window.

**Comment by kenrussell** [comment]
Browsers' compositors are much more involved than a native game drawing directly into a window. The browser's window is opaque; all of the HTML layers that are composited into it, generally, are not.

Taking macOS as a concrete example: here Chrome creates a CoreAnimation layer tree which is rendered into the destination window. WebGL-rendered canvases are promoted into their own layers in the tree and render into an IOSurface which is bound to an OpenGL texture. The IOSurface is handed off to CoreAnimation at the end of each frame. At that level the "opaque" property of CALayer has no effect, so the browser has to make sure that the alpha channel is set to 1.0 beforehand.

Chrome still has code for falling back to doing its own compositing into the destination window, but it's strongly preferred to go through the OS's compositor (on all platforms - Chrome on ChromeOS, macOS, Windows, and Android have support for this) because that achieves about as optimal performance and power consumption as can be had with windowed-mode apps.

It would be both difficult and fragile to make WebGPU-rendered canvases draw directly into the destination window, since this would be incompatible with the entire HTML composition model and could only work in certain special cases. We should focus on making WebGPU work as well as it can with the existing browser composition pipelines.

**Comment by kvark** [comment]
@kenrussell I'm trying to understand... aren't these basically implementation bugs on Apple APIs? I.e. [55509](https://bugs.webkit.org/show_bug.cgi?id=55509) and [200026](https://bugs.webkit.org/show_bug.cgi?id=200026#c3) (thanks @mstange for links!)

For implementation issues, we should expect them fixed and not let them shape the API. Otherwise, nobody is going to fix them :) So if on Apple platforms `premultiplied: false` is slower because we have to do an extra per-multiplication pass, maybe that's ok for the time being until issues are resolved upstream.

**Comment by mstange** [comment]
I wouldn't call the absence of a feature in the OS compositor an implementation bug. The "implementation bug" in the Safari bugs linked above is the fact that Safari does not have a slow-path workaround, unlike Chrome.

I agree with the premise of this bug: There is no efficient way to achieve `alpha: false` or `premultiplied: false` compositing with CoreAnimation, so it is better to leave those values unavailable.
I agree that not having an `alpha: false` option is a bit unfortunate; if CoreAnimation adds an "ignore alpha channel" flag in the future, not having an `alpha: false` option would mean that some theoretical performance is left on the table.

**Comment by kenrussell** [comment]
@kvark glad you followed up and @mstange thanks for confirming and sharing your experience. To be sure:

* none of the OS composition APIs in existence today provide the control that WebGL's `premultipliedAlpha:false` flag does. Definitely not macOS or ChromeOS, and 99% sure not Windows or Android either. This means WebGPU should definitely not expose this control.
* `alpha:false` has no efficient implementation with CoreAnimation today. I'm not sure about ChromeOS and whether it supports RGBX surfaces. Windows might. There is still going to be a large performance cliff on macOS from exposing it. If a future investigation shows that RGBX surfaces are portable among platforms then the WebGPU API can be extended. For now it shouldn't be exposed.

**Comment by RafaelCintron** [comment]
On Windows, DirectComposition gives applications the ability to ignore the alpha channel when composing surfaces together. Doing so allows for better performance, especially in software mode where read-modify-write operations can be replaced with write. 

I am in favor of exposing a way for the web developer to tell the browser to ignore the alpha channel at WebGPU swap chain creation time.

While DirectComposition also gives applications the ability to use "straight" vs. "premultiplied" alpha composition, I think we should keep things simple and standardize on premultiplied for WebGPU.

Is there a way we can influence Apple to allow ignore-alpha surfaces for CoreAnimation in new versions of MacOS?  If so, web developers can at least get good performance moving forward.

**Comment by kenrussell** [comment]
@RafaelCintron I'm reluctant to introduce a piece of functionality into WebGPU which has induced a major performance pitfall in WebGL. Instead I suggest *not* including control over `alpha` for now, and adding it only once it's been demonstrated that all OSs' compositors can support it with no performance overhead.

We can certainly ask our colleagues at Apple to offer this functionality in future macOS releases. It might also be the case that Metal offers better control over this than OpenGL did on macOS. Again, this requires investigation. The expedient path based on our collective experience is to keep the feature set minimal until things are better understood.

**Comment by juj** [comment]
> On Windows, DirectComposition gives applications the ability to ignore the alpha channel when composing surfaces together. Doing so allows for better performance

So Rafael mentions that Windows does support a way to composit without alpha blending enabled.

> @RafaelCintron I'm reluctant to introduce a piece of functionality into WebGPU which has induced a major performance pitfall in WebGL.

Note again that we did not observe the same issues on Firefox or Safari when we profiled alpha:false vs true on TinyRacing in March 2020 when we had a thread about this. The conclusion from that experiment was

```
Enabling alpha does help Chrome, but unfortunately enabling alpha regresses Safari CPU times by +8.2%
(Firefox is unaffected), so it looks like we'll go with a Chrome-specific workaround for this one.
```

We may have mistaken something in that experiment, but those were the results that we saw then. The original test cases are still up at http://clb.confined.space/dump/unity/emunittest_tinyracing/ and  http://clb.confined.space/dump/unity/emunittest_tinyracing.zip to run locally.

It feels a little crazy if `alpha: false` is not supported on OSes, especially on mobile. That would be leaving free performance on the table, alpha blending carries a well known fillrate performance impact. Especially on entry level mobile devices where per pixel performance is really low I would expect it to really make a difference.

Also I don't think it is a good idea to pessimise over some platforms limitations to prevent a meaningful optimization on platforms that are able to do it properly.

If on Windows `alpha: false` is faster than `alpha: true`, and on macOS `alpha: true` is faster than `alpha: false` is fast, picking one as the only supported option could also look a bit like favoritism.

If indeed some platform compositors pass on this "free" optimization, one path forward could be to support all modes, but build an API that enables one to query which blend modes are native, and which ones are emulated, so that applications can discover which will be the fastest path to take.

Not supporting alpha: false and non-PMA alpha does have the performance impact that applications that do need those will need to do extra full frame blits, which would just shift the performance pitfall from browser's side to the application's side. Not sure if that is any improvement. Ideally neither browser side or app side would need to do any extra bs work, but they would be able to agree/discover what the fastest blitting path will be for a given effect.

Maybe such a hypothetical `HTMLCanvasElement.webGpuCompositingMode` setting could have a blend mode `'unused-one'`, which would mean that application does not intend to do alpha blending, but promises to fill the alpha channel with ones. In this mode, the browser could then pick the faster blend implementation, with the guarantee that if the developer followed their end of the bargain, then the browser would provide the proper expected visual result.

**Comment by Kangz** [comment]
I think that just strengthen's @kenrussell's argument that supporting both `alpha:true` and `alpha:false` is complicated and if we do it we should only do it carefully. So for a first release of WebGPU I think it makes sense to only support one of the options and see later if it is worth extending and how it is worth extending. (again in the spirit of getting to a first usable release sooner).

**Comment by magcius** [comment]
There's an important distinction here, which is that WebGL's `{ alpha: false }` is not *just* about compositing, but about the backbuffer format. That is, if you use `{ alpha: false }`, it's supposed to hand you back a `gl.RGB` backbuffer, and there is functionally no alpha, rather than it just being ignored. e.g. if I write linear depth into the alpha of the main render target (as Unreal tends to do), it won't work. This is observable in important circumstances, e.g. blending with DST_ALPHA needs to read back as 1. This is part of the performance hazard.

I don't think anybody wants this behavior. What we *want* is `{ shouldCompositeIntoTheWebPageUsingTheSwapChainAlpha: false }`. In my opinion, we should support a flag like that. Win10's compositor supports RGBX surfaces. Android supports RGBX surfaces. CALayer doesn't support RGBX surfaces, but IMO that's more a limitation of Apple's compositor, and it's relatively easy to fake support for (just draw a full-screen quad that forces alpha back to 1 before presenting).

**Comment by kenrussell** [comment]
Drawing a full-screen quad - or using any other similar technique to fix up the alpha channel - has significant performance implications on GPUs that are fill-rate bound. Specifically, some real-world WebGL content slows down significantly on Intel GPUs if this is done. Please see http://crbug.com/1045643 for exhaustive analysis.

Please agree to accept this limitation in version 1 of the WebGPU API rather than putting in a developer convenience that will introduce per-platform performance issues.

**Comment by magcius** [comment]
Do we have any similar perf analysis for forcing blending during compositing on other platforms? I can't imagine that is similarly cheap -- it will also be slower on fill-rate-bound GPUs. My preference would for Apple to introduce RGBX surface support into CoreAnimation, as RGBX surfaces exist in every other platform composition API I am familiar with.

I'm fine with this limitation for MVP, but perf is being left on the floor in either case. Improving CoreAnimation to support RGBX compositing would help a lot more than just WebGPU.

**Comment by kenrussell** [comment]
Again, we can certainly work with our colleagues at Apple to get this functionality added to CALayer when its contents are set to an IOSurface. We can begin those discussions now - though we'll ideally need a small native test case written using Metal, IOSurface and CALayer mimicking the operations the browser would do.

Appreciate your being willing to move this feature post MVP. I suggest we find some other way to track the feature request with Apple, and close this out so we can reduce the number of open WebGPU investigations.

**Comment by kvark** [comment]
@kenrussell 
> Drawing a full-screen quad - or using any other similar technique to fix up the alpha channel - has significant performance implications on GPUs that are fill-rate bound. Specifically, some real-world WebGL content slows down significantly on Intel GPUs if this is done. Please see http://crbug.com/1045643 for exhaustive analysis.

I think you are confusing the technique WebGL workaround had to use (because of the unfortunate `alpha: false` semantics that @magcius noted) with the technique WebGPU implementation may need to do.

>On Intel GPUs, RGB (rather than RGBA) renderbuffers have poor performance (Issue 607130), so alpha:false is emulated on top of an RGBA back buffer, which is initially cleared to alpha=1.0. Alpha writes are masked via glColorMask(true, true, true, false). This is the only solution I'm aware of that will yield correct blending in all situations, if the user enables various blending modes and attempts to write alpha values less than 1.0 to the framebuffer.

So what you were measuring was very intrusive:
  - clearing alpha to 1 at the beginning of the frame
  - masking all the drawing through the frame with (true, true, true, false)

With WebGPU however, we'd want to do an entirely different thing as a workaround:
  - processing all the draw calls totally unchanged
  - at the end of the pass, draw a full-screen quad setting alpha to 1 with the mask of (false, false, false, true)

The cost of this approach on fill-rate-sensitive systems is exactly 1/4 of the cost of a full-screen quad. It doesn't strike me as unacceptable, especially if we start comparing to the cost of alpha-compositing on systems (i.e. Windows) where users only need opaque compositing.

Interestingly, the bug thread you linked mentions this:
> Clearing the alpha channel explicitly to 1.0 at the end of the frame, using glColorMask(false, false, false, true), reintroduces the stuttering - so this wouldn't be a viable workaround.

I don't understand what this stuttering would be caused by. What we'd be doing is no different from what the user could be doing, there is nothing special to cause stuttering in this approach.

**Comment by Kangz** [comment]
> The cost of this approach on fill-rate-sensitive systems is exactly 1/4 of the cost of a full-screen quad.

Are we sure it's 1/4? I could easily imagine the ROPs loading the texture data to do the write of full texel data. Especially with technique like framebuffer compression, you want to load all the pixel data in the ROPs.

**Comment by kvark** [comment]
@Kangz indeed it could be different from 1/4. My point is - it's not obvious that it's more expensive than forcing the alpha-compositing on platforms that don't have to do this. If we are making the tradeoff, we need to measure these numbers. So far the numbers we have are non-representative (i.e. the WebGL workaround was entirely different).

**Comment by magcius** [comment]
> Are we sure it's 1/4? I could easily imagine the ROPs loading the texture data to do the write of full texel data. Especially with technique like framebuffer compression, you want to load all the pixel data in the ROPs.

This is of-course hardware dependent, and you won't get a clean answer. That said: the point of framebuffer compression is to *reduce* memory bandwidth compared to lossless. It does not make framebuffers smaller in memory (yeah, it's a bit of a confusing name). If framebuffer compression uses *more* memory bandwidth, that means it's not doing its job.

We're asking whether hardware vendors are optimizing for the case of: a fullscreen quad with no pixel shader, color mask = A, blend func = ONE, ONE. That's surprisingly hard to measure, you won't be able to trace that either in Intel GPA or AMD Radeon Analyzer.

**Comment by kenrussell** [comment]
@kvark there is no significant difference between what WebGL and WebGPU would need to do in order to implement `alpha:false`. The dropped frames occurred on GPU-bound content where any additional fill rate consumed by the WebGL implementation pushed the application over the point of being able to reach 60 frames per second. I tried the WebGPU approach in http://crbug.com/1045643 of accumulating, but then ignoring, the alpha channel at the end of the frame. It produced incorrect rendering results, but on the affected content, also used enough fill rate that the app started to drop frames. I think you'll find that `alpha:false` in WebGPU will have a similar cost on this hardware as it did in WebGL.

**Comment by kainino0x** [comment]
@kenrussell There are two different possible semantics, and for WebGPU we're currently considering one that is different from`alpha:false` (and would get a different name):
- Backbuffer is as-if RGB (has no alpha channel - reads always give 1 and writes always no-op) - WebGL, IIUC
- Backbuffer is RGBA (has a normal, writable alpha channel) but the alpha channel is ignored during compositing - WebGPU proposal

**Comment by kainino0x** [comment]
By the way, the latter semantic could probably be backported to WebGL (with the new name) if we decide we like it.

**Comment by kenrussell** [comment]
Understood that there are two different semantics. My high-level point is that the GPU-side work required has a similar performance cost for both approaches, as has been measured with real-world WebGL content in http://crbug.com/1045643 . Still suggest that this feature be deferred until its performance impact can be appropriately characterized on multiple platforms.

**Comment by magcius** [comment]
@kenrussell Am I misreading this? It looks like the end-of-frame clear hasthe same performance cost as forcing `alpha: true` (e.g. what you're suggesting you do in this bug)

> Ordinary alpha:false handling: 39%
>
> Forcing alpha:true (wrong rendering results): 99%
>
> Experiment 1: https://chromium-review.googlesource.com/2285669
> Force alpha mask in ColorMask to false for alpha:false, avoiding setting/resetting ColorMask > around every draw call.
> Result: 38%
>
> Experiment 2: https://chromium-review.googlesource.com/2287369
> Clear the alpha channel at the end of the frame. Attempted to do this both against multisampled renderbuffer and resolved texture; same result. Had rendering artifacts around the piball in this example.
> Result: 99%

**Comment by kenrussell** [comment]
@magcius ah, yes - thanks for double-checking my work. I'd forgotten that the end-of-frame clearing of the alpha channel actually didn't have the same performance problem as the other approaches.

Given that, I withdraw my objection to supporting something like `opaque:true`. Note however that having the alpha channel present, and simply cleared at the end of the frame, will be confusing to some application developers. (It broke the WebGL based pinball game in the abovementioned bug, for example.)

**Comment by kainino0x** [comment]
Editors will make sure we give such an option a very clear name. :)

**Comment by RafaelCintron** [comment]
> Note however that having the alpha channel present, and simply cleared at the end of the frame, will be confusing to some application developers. (It broke the WebGL based pinball game in the abovementioned bug, for example.)

@kenrussell , I do not believe this will be a problem for WebGPU. On Mac, the alpha channel will be cleared after the last web developer instruction runs and the texture is turned over to the user agent.  WebGPU does not have the equivalent of "preserveDrawingBuffer=true".  Even if the texture happens to be reused as a swap chain texture in a subsequent frame, the contents will be lazy cleared so the developer will never see the effects of the alpha clear.

**Comment by RafaelCintron** [comment]
> @kenrussell There are two different possible semantics, and for WebGPU we're currently considering one that is different fromalpha:false (and would get a different name):

@kainino0x , I think we should choose the second semantic you proposed: "Backbuffer is RGBA (has a normal, writable alpha channel) but the alpha channel is ignored during compositing - WebGPU proposal".

**Comment by kenrussell** [comment]
> > Note however that having the alpha channel present, and simply cleared at the end of the frame, will be confusing to some application developers. (It broke the WebGL based pinball game in the abovementioned bug, for example.)
> 
> @kenrussell , I do not believe this will be a problem for WebGPU. On Mac, the alpha channel will be cleared after the last web developer instruction runs and the texture is turned over to the user agent. WebGPU does not have the equivalent of "preserveDrawingBuffer=true". Even if the texture happens to be reused as a swap chain texture in a subsequent frame, the contents will be lazy cleared so the developer will never see the effects of the alpha clear.

@RafaelCintron , the issue with this game is that it enabled blending. There's a semantic difference between the alpha channel always appearing to be 1.0 during the application's rendering, and being updated mid-frame but being cleared to 1.0 at the end of the frame.

The application can adjust its blend functions or shaders to get the desired results. In OpenGL, without masking out the alpha channel in the color mask (which itself led to the performance pitfall), it wasn't possible to change the blend functions or equations underneath the application to give it the appearance of alpha=1.0 all the time.

**Comment by kainino0x** [comment]
Some developers may be surprised by the behavior, but we will definitely well-specify and test it, so it will be up to documentation (and even debugging tools like spector.js) to help users understand the behavior.

Users will then be able to set their blend state accordingly to achieve the result they want.

**Comment by kainino0x** [comment]
I asked some Android experts about this issue. So far, the results are not promising:
[ASurfaceTransaction_setBufferTransparency](https://developer.android.com/ndk/reference/group/native-activity#asurfacetransaction_setbuffertransparency) ([source](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/native/android/surface_control.cpp;l=396;drc=dd7bf2fea566be919c7807b14833c3d3c1fa8458))
[ASURFACE_TRANSACTION_TRANSPARENCY](https://developer.android.com/ndk/reference/group/native-activity#group___native_activity_1gae33c78feb670de33d2abf21ec0624531)

> Updates whether the content for the buffer associated with this surface is completely opaque.
> If true, every pixel of content inside the buffer must be opaque or visual errors can occur.

which sounds like the same semantic as macOS.

**Comment by kainino0x** [comment]
I have a concern about using the clear-alpha-before-compositing strategy on tiled GPUs. We wouldn't be able to do this inside users' render passes (without some serious heroics), which means we potentially have to start a new render pass to do this. (Though perhaps not if underlying APIs have other ways of clearing the alpha channel.) On tilers, adding a whole render pass adds a ton of bandwidth cost.

@kvark had an intriguing idea on chat:

> yeah, that's a bit of an issue.
> I wonder if `storeOp = "present"` could be a thing

I think how this would work is, if the render target is a swap chain texture, and the app submit()s work using `storeOp: "present"`, then we would inject a clear if needed and early-detach the swap chain texture (so it can't be accessed anymore). This probably wouldn't be required, and if it didn't happen the browser would potentially inject a whole extra render pass to clear the alpha channel (and we could warn if this occurs).

**Comment by juj** [comment]
Can this clear-alpha-before-compositing be omitted on platforms that have sane compositing behavior support without performance pitfalls? (on Windows only?)

Would it be possible to put forward a feature request to Android and Apple ecosystems to improve their compositors to avoid the redundant alpha blend when not desired, and needing to do an extra fullscreen GPU pass to guarantee alpha channel data in order to composit?

If a mobile compositor requires one to perform a fullscreen gpu blit pass in order to appease a technicality, one can argue that something has gone wrong in the compositor API design. It would be great to take the message further so that they know that always-on-alpha-blending OS compositors cause real issues. Actually quite a bit surprised to hear that these ecosystems that have had a gaming focus for a decade would not have optimized such limitations out as the first thing? Avoiding doing wasteful alpha blending when not needed is in the "101" types of optimization that GPU vendor slide decks commonly mention towards game developers.

(Is the analysis in this thread accurate? Or is my understanding of the scenario at hand grossly off? Apologies if so)

I would have imagined that low end Android GPUs would certainly avoid all extra GPU work that they can - and such no-benefit alpha blend compositions would be first ones to go. Needing an extra full screen blit pass per frame on top of an already wasted alpha-blend pass is a major perf issue on low end GPUs and on mobile. Web should not be settling with that imo.

**Comment by magcius** [comment]
There's two steps here:

1. The WebGPU implementation will write alpha = 1.0 without any blending.
2. The system compositor will blend window contents with what's under them.

The typical reason that blending is expensive is because of ordering constraints, that is, given two triangles, the first one to be drawn needs to be blended first, meaning that the output of the pixel shader needs to be stored in a temporary buffer until it can blend in the right order. In some cases, these buffers can fill up and new work can't launch in the GPU until these bottlenecks clear up. This is not just draw calls -- overlapping triangles within the same draw call have this ordering constraint.

In both of these two circumstances above, we don't have any ordering constraints. In the clear case, that should be relatively cheap to do on a tiler, since it's just a single triangle that writes a fixed value -- not even a memory load, just a quick write to tiler memory!

And Android has *always* had this system compositor blending cost, so every Android game ever shipped has dealt with it, and it seems to have worked fine for them. WebGPU can't be worse than that. In a lot of cases, the system compositor (hwcomposer on Android) will use display controller planes to do the blending. Think of display controllers as SNES-style hardware that can blend 4 or so positioned sprites, or "planes". The ordering here is well-defined, so there should not be any performance cost from blending in the display controller. In cases it can't do that, it might be the case that the system compositor indeed has to draw a quad with blending.

It would be wonderful to get APIs to get Android to composite an opaque surface. But I don't think this should be that much of a performance hazard -- all we're adding over the usual is a single write to tiler memory.

**Comment by kvark** [comment]
@magcius I don't think you are considering the concern @kainino0x expressed.
Current Android apps can be fine because they can clear alpha to 1 inside the same render pass, so no power is wasted to save to VRAM and read back those colors.
But WebGPU implementation doesn't know, which pass is going to render ti the screen last, so it has to do *an extra* render pass just to clear it. This is where the overhead is (versus native apps). And this is where "storeOp = `present`" could help to narrow down the gap.

**Comment by magcius** [comment]
I was responding to juj about the cost of blending. I don't quite understand the problem with `storeOp: "present"`, but I might want to chat on Matrix about that.

**Comment by kainino0x** [comment]
> Can this clear-alpha-before-compositing be omitted on platforms that have sane compositing behavior support without performance pitfalls? (on Windows only?)

Yes.

> (Is the analysis in this thread accurate? Or is my understanding of the scenario at hand grossly off? Apologies if so)

All correct I think, except IIUC, one major thing: Both macOS and Android allow "please composite this _assuming_ the alpha channel is actually all 1, with undefined results if it's not." So the blending cost can still be saved by the OS.

> And Android has _always_ had this system compositor blending cost,

I'm not sure, but it might be possible to get RGB, no-alpha surfaces from EGL.
